### PR TITLE
Load and render OFF geometry

### LINKED ISSUE #31: Add controls for loading shape from OFF file
Add controls for loading shape from OFF file for all component types.

### LINKED ISSUE #34: Add OFF loading and rendering functionality
Talk with @matthew-d-jones about this, he has some existing Python code which can be adapted.

### PR DESCRIPTION
**Pull request by ACLay**
Closes #31 Closes #34 
image

A selection of OFF files, including the Utah teapot, can be found here: http://www8.cs.umu.se/kurser/TDBC07/HT04/projects/doc/offfiles.html

### CODE REVIEW

File: geometry_constructor/instrument_model.py [hunk 1]
```diff
@@ -53,34 +92,45 @@ def setData(self, index, value, role):
...
             item.translate_vector.z = value
+            self.update_child_transforms(item)
         elif role == InstrumentModel.RotateAxisXRole:
             changed = item.rotate_axis.x != value
             item.rotate_axis.x = value
+            self.update_child_transforms(item)
         elif role == InstrumentModel.RotateAxisYRole:
             changed = item.rotate_axis.y != value
             item.rotate_axis.y = value
+            self.update_child_transforms(item)
```
[the remark this answers is not in the document]
Comment (ACLay) [reply geometry_constructor/instrument_model.py]: Yes it's somewhat repetative, but I'm not sure there'd be a nice helper function for this.

Something like
```
def set_value(self, item, attribute_name, value, transforms=False):
    current_value = getattr(item, attribute_name)
    different = value != current_value
    if different:
        setattr(item, attribute_name, value)
        if transforms:
            self.update_child_transforms(item)
    return different
```
could replace those blocks with 1 line, and would work for some properties, like `item.name` or `item.description`, but not nested ones like `item.translate_vector.x`. The only way I can see to get around that would be to write a separate pair of getter/setter functions for each property, which would defeat the purpose of trying to make the code more concise.

File: requirements.txt [hunk 2]
```diff
@@ -5,3 +5,4 @@ pytest-cov>=2.6.0
 h5py==2.8.0
 flake8==3.5.0
 cx_Freeze==5.1.1
+git+https://github.com/ess-dmsc/python-nexus-utilities@6f5d2c7f1ca66063227aad2a111a2cc567d573f6#egg=nexusutils
```
[the remark this answers is not in the document]
Comment (ACLay) [reply requirements.txt]: That's odd. My machine, and the jenkins build nodes install it as part of the nexusutils install.

When we started this we needed to include a pip update in the jenkins script to sort something... What version of pip is installed in the environment you used?

[the remark this answers is not in the document]
Comment (mattclarke) [reply requirements.txt]: Bleeding edge.
But don't worry about it if it works for you.

File: geometry_constructor/instrument_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mattclarke) [reply geometry_constructor/instrument_model.py]: You only need to do the ones that are repetitive. The special cases can be left as they are.
I haven't time to come up with something but a couple of usual ideas to consider are:
* The pythonic way to deal with a switch statement (which is basically what you have) is often to use a dictionary
* `self.update_child_transform(item)` is called by almost all the 'ifs'. What happens if you call it on an item like InstrumentModel.NameRole?

Perhaps see what you can come up with?

[the remark this answers is not in the document]
Comment (ACLay) [reply geometry_constructor/instrument_model.py]: Turns out that function does work when I just change the arguments 🤦‍♂️
So if I use a dictionary like you suggested to pick the parameters:
```python
    def setData(self, index, value, role):
        row = index.row()
        item = self.components[row]
        changed = False
        param_generators = {
            InstrumentModel.NameRole: lambda: [item, 'name', value, False],
            InstrumentModel.DescriptionRole: lambda: [item, 'description', value, False],
            InstrumentModel.TranslateVectorXRole: lambda: [item.translate_vector, 'x', value, True],
            InstrumentModel.TranslateVectorYRole: lambda: [item.translate_vector, 'y', value, True],
            InstrumentModel.TranslateVectorZRole: lambda: [item.translate_vector, 'z', value, True],
            InstrumentModel.RotateAxisXRole: lambda: [item.rotate_axis, 'x', value, True],
            InstrumentModel.RotateAxisYRole: lambda: [item.rotate_axis, 'y', value, True],
            InstrumentModel.RotateAxisZRole: lambda: [item.rotate_axis, 'z', value, True],
            InstrumentModel.RotateAngleRole: lambda: [item, 'rotate_angle', value, True],
            InstrumentModel.TransformParentIndexRole: lambda: [
                item,
                'transform_parent',
                self.components[value] if value in range(len(self.components)) else None,
                True,
            ],
        }
        if role in param_generators:
            param_list = param_generators[role]()
            changed = self.change_value(*param_list)
        if changed:
            self.dataChanged.emit(index, index, role)
        return changed

    def change_value(self, item, attribute_name, value, transforms):
        current_value = getattr(item, attribute_name)
        different = value != current_value
        if different:
            setattr(item, attribute_name, value)
            if transforms:
                self.update_child_transforms(item)
        return different
```

Does that seem like a better pythonic solution to you?

[the remark this answers is not in the document]
Comment (mattclarke) [reply geometry_constructor/instrument_model.py]: Haha that is really clever!

I was thinking more along the lines of:
```python
...
params = {
            InstrumentModel.NameRole: item.name,
            InstrumentModel.DescriptionRole: item.description,
            ...
            InstrumentModel.RotateAngleRole: item.rotate_angle,
        }

if role in params:
    changed = True if param[role] != value else False
    param[role] = value
    self.update_child_transforms(item)
elif role == InstrumentModel.TransformParentIndexRole:
      # Specialised case
...
```

**Comment by mattclarke** [comment]
Looks good.

**Pull request merged**